File: Iteration2/src/Login/LoginHelper.java
package Login;

import Domain.Course;
import Domain.Department;
import Domain.Lecturer;
import Domain.Student;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Scanner;

public class LoginHelper {
    public void printAvailableCourses(Student student){

        Department department = Department.getInstance();
        Collection<Course> allCourses = department.getAllCourses();
        Collection<Course> availableCourses = new ArrayList<>();
        for (Course course : allCourses){
            if(!student.getTranscript().getCompletedCourses().contains(course)){
                availableCourses.add(course);
            }
        }

        for(Course course : availableCourses){
            System.out.println(course);
        }
    }
    public void listCoursesForLecturer(Department department, Lecturer lecturer) {

        short choice = -1;

        while (choice != 0) {
            System.out.println(lecturer.getCourses());
            System.out.println("\n1 to get action with spesific course");
            System.out.println("0 to go back");
            Scanner input2 = null;
            try{
                input2 = new Scanner(System.in);
                choice = input2.nextShort();
            }catch (Exception e){
                System.out.println(e);
            }
            switch (choice) {
                case 1:
                    System.out.println("Enter course code");
                    String courseCode = null;
                    try{
                        courseCode = input2.next();
                    }catch (Exception e){
                        System.out.println(e);
                        break;
                    }

                    Course course = department.getCourseByCourseCode(courseCode);
                    showCourseOptions(course);
                    break;
                case 0:
                    break;


            }
        }



    }
    public  void showCourseOptions(Course course) {
        if (course == null) {
            System.out.println("Course does not found");
            return;
        }

        short choice = -1;

        while (choice != 0) {
            System.out.println(course.getCourseCode() + "--" + course.getCourseName());
            System.out.println("\n1 to list students");
            System.out.println("2 to grade a student");
            System.out.println("0 to go back");
            Scanner input2 = null;
            try{
                input2 = new Scanner(System.in);
                choice = input2.nextShort();
            }catch (Exception e){
                System.out.println(e);
            }

            switch (choice) {
                case 1:
                    System.out.println(course.getStudents());
                    break;
                case 2:
                    System.out.println("Enter the student number to grade student for course:" + course.getCourseCode());
                    String studentNumber = null;
                    try{
                        studentNumber = input2.next();
                    }catch (Exception e){
                        System.out.println(e);
                        break;
                    }

                    Student student = Department.getInstance().getStudentByStudentNo(studentNumber);
                    if (student == null) {
                        break;
                    }
                    gradeStudent(student, course);
                    showCourseOptions(course);
                    break;
                case 0:
                    choice = 0;
                    break;

            }
        }
    }
    public void gradeStudent(Student student, Course course) {

        System.out.println("Enter a grade for student between 0 and 100");
        float grade = 0;
        try{
            Scanner input = new Scanner(System.in);
            grade = input.nextFloat();
        }catch (Exception e){
            System.out.println(e);
            return;
        }
        System.out.println();
        if (grade < 0 && grade > 100) {
            System.out.println("Invalid grade !!");
            return;
        }

        student.getTranscript().addGrade(course.getCourseCode(), grade);

        System.out.println("Student graded.");
    }


}

File: Iteration3/Login/LoginHelper.py
from Domain.Department import Department
import logging

class LoginHelper:

    def print_available_courses(self, student):
        department = Department.get_instance()
        all_courses = department.get_all_courses()
        available_courses = [course for course in all_courses if
                             course not in student.get_transcript().get_completed_courses()]

        for course in available_courses:
            print(course)

    def list_courses_for_lecturer(self, department, lecturer):
        choice = -1

        while choice != 0:
            for course in lecturer.get_courses():
                print(course)
            print("\n1 to get action with specific course")
            print("0 to go back")

            try:
                choice = int(input())
            except ValueError as e:
                print("You should enter either 0 or 1")
                continue

            if choice == 1:
                print("Enter course code")
                course_code = input()

                course = department.get_course_by_course_code(course_code)
                if course is  None:
                    print(f"Course {course_code} does not found")
                else:
                    self.show_course_options(course)
            elif choice == 0:
                break
            else:
                print("You should enter either 0 or 1")


    def show_course_options(self, course):


        choice = -1

        while choice != 0:
            print(course.get_course_code() + "--" + course.get_course_name())
            print("\n1 to list students")
            print("2 to grade a student")
            print("0 to go back")

            try:
                choice = int(input())
            except ValueError as e:
                print(e)

            if choice == 1:
                for student in course.get_students():
                    print(student.listStudents())
            elif choice == 2:
                print("Enter the student number to grade student for course:" + course.get_course_code())
                student_number = input()

                student = Department.get_instance().get_student_by_student_no(student_number)
                if student is None:
                    break
                self.try_to_grade_student(course, student)
            elif choice == 0:
                break

    def try_to_grade_student(self, course, student):
        if course.get_course_code() in student.get_course_codes():
            self.grade_student(student, course)
        else:
            logger = logging.getLogger()
            logger.info(f"Student with number {student.get_student_no()} couldn't not graded  because  is not taking {course.get_course_code()}")
            print(f"Specified student with number {student.get_student_no()} is not taking {course.get_course_code()}")

    def grade_student(self, student, course):
        print("Enter a grade for student between 0 and 100")
        try:
            grade = float(input())
        except ValueError as e:
            print(e)
            return

        if 0 <= grade <= 100:
            student.get_transcript().add_grade(course.get_course_code(), grade)
            logger = logging.getLogger()
            logger.info(f"Student with number {student.get_student_no()}  graded  for {course.get_course_code()}. Grade is {grade}")

            print("Student graded.")
        else:
            print("Invalid grade!!")
